File: site_generator/benchmarks/map.js
const siteData = require("../site_data.json");

class map {
	constructor() {
		this.stringsmap = {};
	}

	insert(string, url) {
		if (!/^[a-zA-Z]+$/.test(string)) return;
		string = string.toLowerCase();
		if (this.stringsmap[string]) {
			this.stringsmap[string].push(url);
		} else {
			this.stringsmap[string] = [];
			this.stringsmap[string].push(url);
		}
	}

	search(key) {
		if (this.stringsmap[key]) return this.stringsmap[key];
		else return [];
	}
}

mapPtag = new map();

// t1 = Date.now();
// siteData.forEach((site) => {
// 	data = site.title.split(/[^A-Za-z]/);
// 	Object.values(data).forEach((datum) => {
// 		mapHead.insert(datum, site.site_url);
// 	});
// 	data = site.body.split(/[^A-Za-z]/);
// 	Object.values(data).forEach((datum) => {
// 		mapPtag.insert(datum, site.site_url);
// 	});
// });
// t2 = Date.now();
// console.log(t2 - t1);

var k=0, flag=0;
for(var i=0;i<siteData.length;i++){
	site = siteData[i]
	data = site.body.split(/[^A-Za-z]/);
	for(var j=0;j<Object.values(data).length;j++){
		datum = data[j]
		mapPtag.insert(datum, site.site_url);
		k++;
		if(k>=process.argv[2]+100){
			flag=1
			break;
		}
	}
	if(flag==1){
		break
	}
}

t1 = Date.now();
var k=0, flag=0;
for(var i=0;i<siteData.length;i++){
	site = siteData[i]
	data = site.body.split(/[^A-Za-z]/);
	for(var j=0;j<Object.values(data).length;j++){
		datum = data[j]
		mapPtag.search(datum);
		k++;
		if(k>=process.argv[2]){
			flag=1
			break;
		}
	}
	if(flag==1){
		break
	}
}
t2 = Date.now();
console.log(t2 - t1);

File: site_generator/benchmarks/map.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
//	"strconv"
	"strings"
	"time"
)

type Site struct {
	Title      string   `json:"title"`
	Body       string   `json:"body"`
	Links      []int    `json:"links"`
	ParentSite string   `json:"parent_site"`
	SiteURL    string   `json:"site_url"`
	LinkURL    []string `json:"link_urls"`
	Slug       string   `json:"slug"`
}

const (
	ALBHABET_SIZE = 26
)

type trieNode struct {
	childrens [ALBHABET_SIZE]*trieNode
	isWordEnd bool
}

type trie struct {
	root *trieNode
}

func initTrie() *trie {
	return &trie{
		root: &trieNode{},
	}
}

func (t *trie) insert(word string) {
	wordLength := len(word)
	current := t.root
	for i := 0; i < wordLength; i++ {
		index := word[i] - 'a'
		if current.childrens[index] == nil {
			current.childrens[index] = &trieNode{}
		}
		current = current.childrens[index]
	}
	current.isWordEnd = true
}

func (t *trie) find(word string) bool {
	wordLength := len(word)
	current := t.root
	for i := 0; i < wordLength; i++ {
		index := word[i] - 'a'
		if current.childrens[index] == nil {
			return false
		}
		current = current.childrens[index]
	}
	if current.isWordEnd {
		return true
	}
	return false
}

func main() {
	jsonFile, err := os.Open("../site_data.json")
	if err != nil {
		fmt.Println(err)
	}

	defer jsonFile.Close()
	byteValue, _ := ioutil.ReadAll(jsonFile)
	var sites []Site
	json.Unmarshal(byteValue, &sites)

	trie := initTrie()

	t1 := time.Now()

	for i := 0; i < len(sites); i++ {
		sentence := strings.Split(sites[i].Body, " ")
		for j := 0; j < len(sentence); j++ {
			sentence[j] = strings.ToLower(sentence[j])
			matched, _ := regexp.MatchString(`^[a-z]*$`, sentence[j])
			// fmt.Println(matched, sentence[j])
			if matched {
				trie.insert(sentence[j])
			}
		}
	}

//	t1 := time.Now()
/*	no_of_words, err := strconv.Atoi(os.Args[1])
	k := 0
	fl := 0
	for i := 0; i < len(sites); i++ {
		sentence := strings.Split(sites[i].Body, " ")
		for j := 0; j < len(sentence); j++ {
			sentence[j] = strings.ToLower(sentence[j])
			matched, _ := regexp.MatchString(`^[a-z]*$`, sentence[j])
			if matched {
				trie.find(sentence[j])
				k++
				if k >= no_of_words {
					fl = 1
					break
				}
			}
		}
		if fl == 1 {
			break
		}
	}*/
	t2 := time.Now()
	elapsed := t2.Sub(t1)
	fmt.Println(elapsed)
}
